src/lsp.ts

```diff
@@ -0,0 +1,58 @@
+import * as which from "which";
+import {
+  findLanguageServer,
+  setupLanguageServer,
+} from "./utils";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { LanguageClient } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration
+  ls: LanguageClient | null = null
+  executableName: string
+  context: ExtensionContext
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName
+    this.context = context
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  abstract startLanguageServer(serverModule: string): void
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  canDownloadLanguageServer(): boolean {
+    return this.downloadInfo != null && findLanguageServer(this) == null
+  }
+
+  get languageServerPath(): string | null {
+    return this.config.languageServerPath || findLanguageServer(this) || which.sync(this.executableName, { nothrow: true })
+  }
+
+  async setupLanguageServer() {
+    setupLanguageServer(this)
```

Comment:
Seems like you should just inline the other function to me. No need to call an out of band function.